cardano-api/src/Cardano/Api/Typed.hs

```diff
@@ -1561,111 +1562,129 @@ data MultiSigScript = RequireSignature (Hash PaymentKey)
+                                      case toBoundedInteger sci of
+                                        Just reqInt ->
+                                          Just $ fromJSONmOfnChecks
+                                                   (gatherMultiSigScripts listVectorValue)
+                                                   reqInt
+                                                   (Vector.length listVectorValue)
+                                        Nothing -> error $ "Error in multisig \"required\" key: "
+                                                         <> show sci <> " is not a valid Int"
+                                   (_, _) -> Nothing
+                     Just mss -> Just mss
+                 _ -> Nothing
+
+  -- To account for backwards compatibility
+  previousScriptSyntax :: HMS.HashMap Text Value -> Maybe MultiSigScript
+  previousScriptSyntax hm =
+        case HMS.lookup "paymentKeyHashes" hm of
+          Just (Array pkhListVectorValue) ->
+            case HMS.lookup "required" hm of
+              Nothing ->
+                error "atLeast multisig missing \"required\" key"
+              Just (Number sci) ->
+                case toBoundedInteger sci of
+                  Just reqInt ->
+                    Just $ fromJSONmOfnChecks
+                             (gatherMultiSigScripts pkhListVectorValue)
+                             reqInt
+                             (Vector.length pkhListVectorValue)
+                  Nothing -> error $ "Error in multisig \"required\" key: "
+                                   <> show sci <> " is not a valid Int"
+              _ -> Nothing
+          _ -> Nothing
+
+  fromJSONmOfnChecks :: [MultiSigScript] -> Int -> Int -> MultiSigScript
+  fromJSONmOfnChecks keys required total
+    | total <= 0 = error $ "No payment key hashes were parsed from multiscript object: " <> show keys
+    | required <= 0 = error "The required number of payment key hashes cannot be less than or equal to 0."
+    | required == 1 = error "required is equal to one, you should use the \"any\" multisig script"
+    | length keys < required = error $ "required exceeds the number of payment key hashes. \
+                                       \Number of keys: " ++ show (length keys)
+                                     ++ " required: " ++ show required
+    | length keys == total = RequireMOf required keys
+    | otherwise = error $ "Cardano.Api.Typed.fromJSONmOfnChecks failure occured. required: "
+                        ++ show required ++ " total: " ++ show total
+                        ++ " keys: " ++ show keys
+
+  convertToHash :: Text -> Hash PaymentKey
+  convertToHash txt = case deserialiseFromRawBytesHex (AsHash AsPaymentKey) $ Text.encodeUtf8 txt of
+                        Just payKeyHash -> payKeyHash
+                        Nothing -> error $ "Error deserialising payment key hash: " <> Text.unpack txt
```

Comment:
Maybe we should consider using `error` less because it means users of the API will be unable to handle this error programmatically (except maybe in `IO` by catching an exception).

A fix would be returning `Either Text (Hash PaymentKey)` instead.

Reply:
In this instance I think its ok because it's at the top level close to the JSON parsing.